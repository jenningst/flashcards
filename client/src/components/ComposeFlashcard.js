import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { PrimaryButton } from './elements/Button';

function ComposeFlashcard({
  handleTextChange,
  handleAnswerChange,
  questionText,
  questionAnswer
}) {
  const [showAnswer, setShowAnswer] = useState(false); 
  const toggleAnswer = () => setShowAnswer(!showAnswer);

    return (
    <ComposeFlashcardWrapper className="ComposeFlashcard">
      <QuestionTextInput className="ComposeFlashcard__input">
        {!showAnswer
          ? <textarea 
              className="ComposeFlashcard__input--text"
              data-testid="question-text"
              onChange={handleTextChange}
              value={questionText}
              placeholder="Type your question here..."
            />
          : <textarea
              className="ComposeFlashcard__input--answer"
              data-testid="question-answer"
              onChange={handleAnswerChange}
              value={questionAnswer}
              placeholder="Type your answer here..."
            />
        }
      </QuestionTextInput>
      <ToggleButton
        className="ComposeFlashcard__button-toggle"
        type="button"
        onClick={toggleAnswer}
      >
        {!showAnswer ? 'SHOW ANSWER' : 'SHOW QUESTION'}
      </ToggleButton>
    </ComposeFlashcardWrapper>
  );
};

const ComposeFlashcardWrapper = styled.div`
  box-sizing: border-box;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  height: 100%;
  padding: 1.5rem;

  border-bottom-left-radius: 1rem;
  border-bottom-right-radius: 1rem;
  box-shadow: 0px 10px 18px -11px rgba(120,119,120,1);
  background: ${props => props.theme.color.main.pureWhite};
  color: ${props => props.theme.color.fonts.charleston};
`;

const QuestionTextInput = styled.div`
  flex-grow: 1;
  width: 100%;
  margin-bottom: 1rem;

  textarea {
    box-sizing: border-box;
    width: 100%;
    height: 100%;
    padding: .50rem;
    resize: none;
    font-family: 'Rubik', sans-serif;
    font-size: 15px;
    font-weight: 300;
    line-height: 20px;
    outline: 1px dashed ${props => props.theme.color.fonts.lightGrey};

    :focus {
      background-color: ${props => props.theme.color.fonts.lightGrey};
      outline: 1px dashed ${props => props.theme.color.fonts.charleston};
      border: none;
    }

    ::placeholder {
      color: ${props => props.theme.color.fonts.lightGrey};
      font-family: 'Rubik', sans-serif;
      font-size: 15px;
      font-weight: 300;
      line-height: 20px;
    }
  }
`;

const ToggleButton = styled(PrimaryButton)`
  width: 100%;
  background: ${props => props.theme.color.main.primary};
  border-bottom: 3px solid ${props => props.theme.color.main.primaryHover};
  border-radius: 15px;
  color: ${props => props.theme.color.fonts.pureWhite};
  outline: none;
  box-shadow: 0px 10px 18px -11px rgba(120,119,120,1);

  &:hover {
    background: ${props => props.theme.color.main.primaryHover};
    border-bottom: 3px solid ${props => props.theme.color.main.primary};
  }
  
  &:active {
    background: ${props => props.theme.color.main.secondaryHover};
  }
`;

ComposeFlashcard.propTypes = {
  handleTextChange: PropTypes.func.isRequired,
  handleAnswerChange: PropTypes.func.isRequired,
  questionText: PropTypes.string,
  questionAnswer: PropTypes.string,
};

ComposeFlashcard.defaultProps = {
  questionText: '',
  questionAnswer: '',
}

export default ComposeFlashcard;

