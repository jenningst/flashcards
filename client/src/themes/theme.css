/* 

button: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      active: '#5860A5',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    border: {
      primary: '#42487C',
      active: '#939CE8'
      disabled: '#b1aeb7',
      input: '#e9eaf0',
    }, // TODO: consolidate everything below; using the above
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }

}*/