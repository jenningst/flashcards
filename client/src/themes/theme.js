export const darkTheme = {
  font: {
    primary: '#FFFFFF',
    secondary: '#E0E0E0',
    tertiary: '#666666',
  },
  background: {
    primary: '#282626',
    secondary: '#313131',
    highlight: '#6121BF',
    special: '#7358F7',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};

export const lightTheme = {
  color: {
    button: {
      primary: '#939CE8',
      primaryHover: '#7883E2',
      active: '#5860A5',
      secondary: '#50cdff',
      secondaryHover: '#6FD6FF',
      disabled: '#e0e0eb',
    },
    font: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      eerieBlack: '#16171D',
      charleston: '#33364d',
      blackCoral: '#585A6D',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1',
      placeholder: '#e9eaf0',
      danger: '#DE6A60',
    },
    border: {
      primary: '#5860A5',
      secondary: '#7883E2',
      disabled: '#b1aeb7',
      inputDefault: '#e9eaf0',
      transparent: '#0000ffff',
      eerieBlack: '#16171D',
      charleston: '#33364d',
    }, // TODO: consolidate everything below; using the above
    background: {
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
    },
    main: {
      primary: '#7883E2',
      primaryHover: '#939CE8',
      secondary: '#e3e5ff',
      secondaryHover: '#c3d2ff',
      pureWhite: '#ffffff',
      offWhite: '#f5f7fb',
    },
    alerts: {
      notice: '#c0a5ff',
      success: '#50cdff',
      successHover: '#6FD6FF',
      warning: '#FFCB17',
      warningBg: '#FFDA5C',
      danger: '#DE6A60',
    },
    fonts: {
      eerieBlack: '#16171D',
      charleston: '#292b3d',
      pureWhite: '#ffffff',
      offWhite: '#fcfbfc',
      grey: '#b1aeb7',
      lightGrey: '#eff0f1'
    },
    borders: {
      lightGrey: '#d8d6db',
    }
  },
  spacing: {
    compact: '1rem',
    comfortable: '1.5rem',
    generous: '2rem',
  },
  font: {
    primary: '#4e5167',
    secondary: '#262626',
    tertiary: '#666666',
  },
  background: {
    primary: '#ffffff',
    secondary: '#fcfbfc',
    highlight: '#9388ff',

    special: '#50cdff',
    inactive: '#b1aeb7',
  },
  button: {
    default: {
      primary: '#7358F7',
      border: '#4D18AE',
      alert: '#D55949',
      greyed: '#666666',
    },
    disabled: {
      primary: '',
      secondary: '',
      border: '',
    },
    hover: {
      primary: '#6121BF',
      border: '#6121BF',
    },
    active: {
      primary: '#8b90ff',
      border: '',
    },
  },
};