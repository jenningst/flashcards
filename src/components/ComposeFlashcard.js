import React, { useState, useContext, useRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import ThemeContext from '../contexts/themeContext';

import { Body } from '../components/Elements/Text';
import { MediumButton } from '../components/Elements/Button';
import { ReactComponent as EditIcon } from '../'

ComposeFlashcard.propTypes = {

};

function ComposeFlashcard({ question }) {
  const theme = useContext(ThemeContext);
  const [showAnswer, setShowAnswer] = useState(false);
  const [isEditable, setIsEditable] = useState(true);
  const [questionText, setQuestionText] = useState('');
  const [questionAnswer, setQuestionAnswer] = useState('');

  const handleTextChange = e => setQuestionText(e.target.value);
  const handleAnswerChange = e => setQuestionAnswer(e.target.value);
  const toggleAnswer = () => {
    setShowAnswer(!showAnswer);
  };
  const toggleEditable = () => setIsEditable(!isEditable);

    return (
    <ComposeFlashcardWrapper className="ComposeFlashcard">
      <div className="ComposeFlashcard__question">
        {!showAnswer
          ? <textarea 
              className="ComposeFlashcard__input--text"
              onChange={handleTextChange}
              value={questionText}
            />
          : <textarea
              className="ComposeFlashcard__input--answer"
              onChange={handleAnswerChange}
              value={questionAnswer}
            />
        }
      </div>
      <MediumButton
        className="ComposeFlashcard__toggle-button"
        type="button"
        onClick={toggleAnswer}
      >
        {!showAnswer ? 'SHOW ANSWER' : 'SHOW QUESTION'}
      </MediumButton>
    </ComposeFlashcardWrapper>
  );
};

const ComposeFlashcardWrapper = styled.div`
  box-sizing: border-box;
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;
  height: 100%;

  border-radius: 20px;
  margin-left: 1rem;
  margin-right: 1rem;
  padding: 1.5rem 1rem 1rem 1rem;
  border: solid 1px grey;

  .ComposeFlashcard__question {
    flex-grow: 1;
    width: 100%;
    margin-bottom: 1rem;
    
    .ComposeFlashcard__question-text {
      margin: 0;
    }

    textarea[class^="ComposeFlashcard__input"] {
      box-sizing: border-box;
      height: 100%;
      width: 100%;
      border: none;
      resize: none;

      font-family: 'Rubik', sans-serif;
      font-size: 15px;
    }
  }

  .ComposeFlashcard__toggle-button {
    outline: none;
    &:hover {
    }
    &:active {
    }
    border-radius: 15px;
    width: 100%;
  }
`;

export default ComposeFlashcard;

// const ComposeFlashcardWrapper = styled.div`
//     box-sizing: border-box;
//     display: flex;
//     flex-flow: column nowrap;
//     justify-content: center;
//     align-items: center;
//     height: 100%;
  
//     border-radius: 20px;
//     margin-left: 1rem;
//     margin-right: 1rem;
//     padding: 1.5rem 1rem 1rem 1rem;
//     background: red;
//     color: ${theme.font.secondary};
  
//     .ComposeFlashcard__question {
//       flex-grow: 1;
//       width: 100%;
      
//       .ComposeFlashcard__question-text {
//         margin: 0;
//       }
//     }
  
//     .ComposeFlashcard__toggle-button {
//       background: ${theme.button.default.primary};
//       border: 2px solid ${theme.button.default.primary};
//       color: ${theme.font.primary};
//       outline: none;
//       &:hover {
//         background: ${theme.button.hover.primary};
//         border: 2px solid ${theme.button.hover.primary};
//       }
//       &:active {
//         border: 2px solid ${theme.button.active.primary};
//       }
//       border-radius: 15px;
//       width: 100%;
//     }
//   `;