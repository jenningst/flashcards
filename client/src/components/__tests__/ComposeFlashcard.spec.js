import React from 'react';
import { ThemeProvider } from 'styled-components';
import { darkTheme, lightTheme } from '../../themes/theme';

import {
  cleanup,
  fireEvent,
  render,
} from '@testing-library/react';

import ComposeFlashcard from '../ComposeFlashcard';

afterEach(cleanup);

const renderComponent = ({
  handleTextChange,
  handleAnswerChange ,
  questionText,
  questionAnswer,
  theme
}) =>
  render(
    <ThemeProvider theme={theme}>
      <ComposeFlashcard
        handleTextChange={handleTextChange}
        handleAnswerChange={handleAnswerChange}
        questionText={questionText}
        questionAnswer={questionAnswer}
      />
    </ThemeProvider>
  );

describe('<ComposeFlashcard /> spec', () => {
  it('assert component matches snapshot', () => {
    const { container } = renderComponent({
      handleTextChange: jest.fn(),
      handleAnswerChange: jest.fn(),
      questionText: '',
      questionAnswer: '',
      theme: lightTheme,
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  // TODO: can we refactor test setup into a better helper function?

  it('assert initially renders an empty question text textarea', () => {
    // given: the component has initially rendered
    // when: it mounts
    // then: the user should see an empty text area for the question text
    const textChange = jest.fn();
    const answerChange = jest.fn();
    const { getByTestId } = renderComponent({
      handleTextChange: textChange,
      handleAnswerChange: answerChange,
      questionText: '',
      questionAnswer: '',
      theme: lightTheme,
    });

    expect(getByTestId('question-text')).toBeEmpty();
  });

  it('assert a toggle button is initially rendered with SHOW ANSWER', () => {
    // given: the component has initially rendered
    // when: it mounts
    // then: there should be a toggle button with text 'SHOW ANSWER'
    const textChange = jest.fn();
    const answerChange = jest.fn();
    const { getByText } = renderComponent({
      handleTextChange: textChange,
      handleAnswerChange: answerChange,
      questionText: '',
      questionAnswer: '',
      theme: lightTheme,
    });

    expect(getByText('SHOW ANSWER')).toBeEnabled();
  });

  it('assert handleTextChange handler is called', async () => {
    // given: the text area after initial render
    // when: a user types into the field
    // then: the handleTextChange handler should be called
    const textSpy = jest.fn();
    const answerSpy = jest.fn();
    const newText = 'What do you call someone with no body and no nose?';
    const mockEvent = { target: { value: newText }};
    const { getByTestId } = renderComponent({
      handleTextChange: textSpy,
      handleAnswerChange: answerSpy,
      questionText: '',
      questionAnswer: '',
      theme: lightTheme,
    });
    const textAreaElement = getByTestId('question-text');

    fireEvent.change(textAreaElement, mockEvent);

    expect(textSpy).toHaveBeenCalledTimes(1);
  });
  
  it('assert initially renders an empty answer text textarea', async () => {
    // given: the question text area has valid input
    // when: a user toggles to the answer
    // then: the toggle button text should change
    const textSpy = jest.fn();
    const answerSpy = jest.fn();
    const newText = 'Did you know the first French fries weren\'t actually cooked in France?';
    const { getByText, findByTestId } = renderComponent({
      handleTextChange: textSpy,
      handleAnswerChange: answerSpy,
      questionText: newText,
      questionAnswer: '',
      theme: lightTheme,
    });

    fireEvent.click(getByText(/^show answer$/i), { button: 0 });

    expect(await findByTestId('question-answer')).toBeEmpty();
  });

  it('assert toggle button text changes upon valid input', async () => {
    // given: the question text area has valid input
    // when: a user toggles to the answer
    // then: the toggle button text should change
    const textSpy = jest.fn();
    const answerSpy = jest.fn();
    const newText = 'Did you know the first French fries weren\'t actually cooked in France?';
    const { getByText, findByText } = renderComponent({
      handleTextChange: textSpy,
      handleAnswerChange: answerSpy,
      questionText: newText,
      questionAnswer: '',
      theme: lightTheme,
    });

    fireEvent.click(getByText(/^show answer$/i), { button: 0 });
    expect(await findByText(/^show question$/i)).toBeEnabled();

  });

  it('assert handleAnswerChange handler is called', async () => {
    // given: the text area after initial render
    // when: a user types into the field
    // then: the handleTextChange handler should be called
    const textSpy = jest.fn();
    const answerSpy = jest.fn();
    const newText = 'What do you call someone with no body and no nose?';
    const newAnswer = 'Nobody knows.'
    const mockEvent = { target: { value: newAnswer }};
    const { getByText, findByTestId } = renderComponent({
      handleTextChange: textSpy,
      handleAnswerChange: answerSpy,
      questionText: newText,
      questionAnswer: '',
      theme: lightTheme,
    });

    fireEvent.click(getByText(/^show answer$/i), { button: 0 });
    fireEvent.change(await findByTestId('question-answer'), mockEvent);

    expect(answerSpy).toHaveBeenCalledTimes(1);
  });

  it('assert question text persists after toggle', async () => {
    // given: the question text and question answer textareas have valid input
    // when: a user toggles from the answer to the question text
    // then: the question text should persist
    const textSpy = jest.fn();
    const answerSpy = jest.fn();
    const newText = 'How do you make holy water?';
    const newAnswer = 'You boil the hell out of it.'
    const { getByText, findByTestId } = renderComponent({
      handleTextChange: textSpy,
      handleAnswerChange: answerSpy,
      questionText: newText,
      questionAnswer: '',
      theme: lightTheme,
    });

    fireEvent.click(getByText(/^show answer$/i), { button: 0 });
    expect(await findByTestId('question-answer')).toBeInTheDocument();

    fireEvent.click(getByText(/^show question$/i), { button: 0 });
    expect(await findByTestId('question-text')).toBeInTheDocument();
    expect(await findByTestId('question-text')).toHaveValue(newText);
  });
});