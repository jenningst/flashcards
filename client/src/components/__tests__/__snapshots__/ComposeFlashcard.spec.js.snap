// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ComposeFlashcard /> spec assert component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ComposeFlashcard sc-bxivhb brNQUz"
  >
    <div
      class="ComposeFlashcard__input sc-ifAKCX jbQeev"
    >
      <textarea
        class="ComposeFlashcard__input--text"
        data-testid="question-text"
      />
    </div>
    <button
      class="sc-bwzfXH ComposeFlashcard__button-toggle sc-EHOje dYtATZ"
      type="button"
    >
      SHOW ANSWER
    </button>
  </div>
</DocumentFragment>
`;
